pkg/funky/router.go

```diff
@@ -0,0 +1,170 @@
+///////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+package funky
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// FirstPort the starting port number for servers created by a Router
+const FirstPort uint16 = 9000
+
+// Healthy a channel for reporting the health of the web service.
+var Healthy = make(chan struct{})
+
+// Router an interface for delegating function invocations to idle servers
+type Router interface {
+	Delegate(input *Message) (*Message, error)
+	Shutdown() error
+}
+
+// DefaultRouter a struct that hold servers that can be delegated to
+type DefaultRouter struct {
+	servers       []Server
+	serverFactory ServerFactory
+	mutex         *sync.Mutex
+	sem           *semaphore.Weighted
+}
+
+// NewRouter constructor for DefaultRouters
+func NewRouter(numServers int, serverFactory ServerFactory) (*DefaultRouter, error) {
+	if numServers < 1 {
+		return nil, IllegalArgumentError("numServers")
+	}
+
+	servers, err := createServers(numServers, serverFactory)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DefaultRouter{
+		servers:       servers,
+		serverFactory: serverFactory,
+		mutex:         &sync.Mutex{},
+		sem:           semaphore.NewWeighted(int64(numServers)),
+	}, nil
+}
+
+// Delegate delegates function invocation to an idle server
+func (r *DefaultRouter) Delegate(input *Message) (*Message, error) {
+	server, err := r.findFreeServer()
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		r.releaseServer(server)
+	}()
+
+	var e Error
+	resp, err := server.Invoke(input)
+
+	logs := Logs{
+		Stdout: server.Stdout(),
+		Stderr: server.Stderr(),
+	}
+
+	if err != nil {
+		switch v := err.(type) {
+		case TimeoutError:
+			defer func() {
+				recover()
+			}()
+			terminateErr := server.Terminate()
```

Comment:
We should set `server = nil` here, because we don't want to release the terminated server back into the pool